File: 백준/Silver/18352. 특정 거리의 도시 찾기/특정 거리의 도시 찾기.py
import sys
from collections import deque

input = sys.stdin.readline

V, E, K, start = map(int, input().split())
adj = [[] for _ in range(V + 1)]
for _ in range(E):
    a, b = map(int, input().split())
    adj[a].append(b)

visited = [False] * (V + 1)
distance = [0] * (V + 1)


def bfs(start):
    answer_list = []
    que = deque([start])
    visited[start] = True
    distance[start] = 0
    while que:
        now = que.popleft()
        for i in adj[now]:
            if not visited[i]:
                visited[i] = True
                que.append(i)
                distance[i] = distance[now] + 1
                if distance[i] == K:
                    answer_list.append(i)

    if len(answer_list) == 0:
        print(-1)
    else:
        answer_list.sort()
        for i in answer_list:
            print(i)


bfs(start)

File: 백준/Silver/18352. 특정 거리의 도시 찾기/특정 거리의 도시 찾기.js
// const input = require('fs')
//   .readFileSync('18352.txt')
//   .toString()
//   .trim()
//   .split('\n');
const input = require('fs')
  .readFileSync('dev/stdin')
  .toString()
  .trim()
  .split('\n');
const [N, M, K, X] = input.shift().split(' ').map(Number);
const arr = input.map((v) => v.split(' ').map(Number));
const graph = [...Array(N + 1)].map(() => []);
const distance = Array(N + 1).fill(0);
let answer = [];

arr.map(([from, to]) => graph[from].push(to));

const bfs = (start) => {
  const queue = [start];
  distance[start] = 1;

  while (queue.length) {
    const now = queue.shift();
    if (distance[now] === K + 1) {
      answer.push(now);
      continue;
    }
    for (const next of graph[now]) {
      if (distance[next] === 0) {
        queue.push(next);
        distance[next] = distance[now] + 1;
      }
    }
  }
};

bfs(X);
if (answer.length) {
  answer.sort((a, b) => a - b);
  console.log(answer.join('\n'));
} else {
  console.log(-1);
}
